### ISSUE TITLE
Support reading a graphviz DOT file.

### ISSUE THREAD

**Issue by gaborcsardi** [igraph/igraph#447]
Last time I checked the graphviz licence did not allow using its DOT parser in igraph. :(

---

Imported from Launchpad using lp2gh.
- date created: 2012-01-10T13:50:42Z
- owner: gabor.csardi
- the launchpad url was https://bugs.launchpad.net/bugs/914275

**Comment by dirmeier** [comment]
Hello,

are there any updates on this? The parser from [pydot](https://github.com/erocarrera/pydot/blob/master/dot_parser.py) looks as if it is the official one and is also used in `networkx` (it's a MIT License). It would be great to be able to directly read dot files with `igraph`.

Best,
Simon

**Comment by ntamas** [comment]
Unfortunately even if the one in pydot is functionally equivalent to the "official" parser, we would still need to implement an equivalent parser in C (preferably based on Flex/Bison as it is used for all the other parsers in the C core). We are happy to accept pull requests for such a parser implementation, but it's unlikely that we will have time to implement it ourselves in the near future.

**Comment by dirmeier** [comment]
Would it suffice to implement the respective method in the src here:
https://github.com/igraph/igraph/blob/674855ac2e2aae00c8eb670c8483c9bf6e4ffe25/src/foreign.c#L23
and provide a respective header like this one
https://github.com/igraph/igraph/blob/674855ac2e2aae00c8eb670c8483c9bf6e4ffe25/src/foreign-ncol-header.h#L23
If so would I need to provide a lexer and parser, too?

**Comment by ntamas** [comment]
Yes; most of the reader functions implemented in igraph use some kind of an internal data structure (like the one you have found in `foreign-ncol-header.h`) to collect all the information parsed out from the input file. The generated lexer and parser manipulates this data structure. Then, once everything is parsed, a few final calls to `igraph_add_vertices()` and `igraph_add_edges()` are performed to build the graph itself.

**Comment by szhorvat** [comment]
I will leave this open, but I think we should not implement this, because:

 - DOT is the native format for Graphviz, tailored to the specificities of that system. It is not a format designed for general purpose graph exchange and archival.
 - The format is quite complex and can store many things that igraph doesn't support.
 - Implementing format readers is very expensive and time consuming. We want readers that are robust enough to withstand the scrutiny of OSS-fuzz, and won't crash on bad input.
 - Even though there's a formal spec on the Graphviz website, [this is not accurate](https://forum.graphviz.org/t/grammaar-in-document-icorrect/976).
